Add healthz check for KMS Providers on kube-apiserver.

immutableT opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
When kube-apiserver starts before kms-plugin, kube-apiserver essentially is unable to serve secrets until kms-plugin is up. This inability to process secrets (just after being started) leads to large number of errors and eventually to a restart of kube-apiserver. 
To avoid this untested scenario, KMS Provider will install a healthz check on kube-apiserver for the status of kms-plugin.
Therefore, (via a readiness probe) kube-apiserver will be protected from serving requests until kms-plugin becomes available. Thus allowing kube-apiserver properly handle requests for Secrets and Service Accounts.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
KMS Providers will install a healthz check for the status of kms-pluign in kube-apiservers' encryption config. 
```

Commit:
Allow kube-apiserver to test the status of kms-plugin.

immutableT commented:
/assign @awly

immutableT commented:
/test pull-kubernetes-bazel-build

immutableT commented:
/test pull-kubernetes-bazel-test

awly commented on staging/src/k8s.io/apiserver/pkg/apis/config/types.go:
```diff
@@ -89,4 +89,13 @@ type KMSConfiguration struct {
 	Timeout *metav1.Duration
+	// HealthzURL is the url on which kube-apiserver could check the health status of kms-plugin.
+	// +optional
+	HealthzURL string
```
Could healthz probes be passed over the plugin `Endpoint` above?

immutableT replied:
I assume that you meant to create a dedicated type for healthz probe. If I understood you correctly, the latest commit accomplishes this.

awly replied:
I meant adding a healthz RPC to the KMS plugin itself on the gRPC service.
So you could send those healthz requests over KMSConfiguration.Endpoint (the unix socket) instead of a separate new HTTP endpoint

immutableT replied:
I think this is a good idea, but my survey of the currently implemented plugins shows that they do not have this functionality yet.
This is something we could add later.

awly replied:
Do currently implemented plugins already expose a HTTP endpoint for healthz?
If not, you could put in a healthz grpc endpoint and just skip polling health if it returns "unimplemented" (or whatever error grpc will send).

immutableT replied:
Yes, Google CloudKMS Plugin and AWS CloudHSM plugins have implemented healthz. Hence, these plugins could take advantage of this feature right away.

awly replied:
Ack, that makes sense then

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -20,15 +20,20 @@ import (
 	"time"
 
+	"k8s.io/klog"
+
```
remove empty line

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -309,6 +322,17 @@ func GetSecretboxPrefixTransformer(config *apiserverconfig.SecretboxConfiguratio
 func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, envelopeService envelope.Service, prefix string) (value.PrefixTransformer, error) {
+	if config.HealthzURL != "" {
+		klog.V(4).Infof("Validating healthz of KMS-Plugin via %s", config.HealthzURL)
+		ready, err := isKMSPluginReady(config)
```
So this only happens on kube-apiserver startup, right?
What if kms plugin becomes unhealthy later on? I assume it will just show up as errors from all Secrets operations?

immutableT replied:
Yes, this is only to ensure that kube-apiserver comes-up after kms-plugin is reporting healthz OK.
KMS-Plugin may have its own liveness probe.
And yes, kube-apiserver will log and increment failure metrics when the plugin is down.

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +342,62 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
 	}, nil
 }
+
+func isKMSPluginReady(config *apiserverconfig.KMSConfiguration) (bool, error) {
```
Could this only return the error? If arg parsing fails _or_ if healthz polling fails, return a non-nil error

immutableT replied:
Maybe, but I want to deffirentate between the scenarios 
1. Healthz could not be checked due to configuration errors - false, error
2. Heathz returned non 200 status - false,nil
3. Healthz return 200 - true,nil

I think this approach fits with the signature of the poll.Wait.

awly replied:
Hmm, but shouldn't you retry even if healthz endpoint is unreachable?
For example when KMS plugin starts later than kube-apiserver.
If yes, then there's no difference between non-200 status and any request errors

immutableT replied:
Let me step back.
wait.PollImmediate takes a condition function of the following signature:
type ConditionFunc func() (done bool, err error)
Therefore, getHelathz must match that signature.

I think the reason for that signature is that polling status has three potential states:
1. true - we are done ,no need to retry further (kms-plugin OK)
2. false - poll failed, keep retrying (kms-plugin is not up yet)
3. error - got negative status, no need to poll anymore (ex. missing permissions on the key)

awly replied:
`getHealthz` doesn't have to conform to what `wait.PollImmediate` expects, you can always adapt it with a closure.

I understand the purpose of those separate return values, but my point is: getHealthz should never cause wait.PollImmediate to exit early with an error.
Currently you return non-nil error when response code is not 200. I think kube-apiserver should keep retrying in that case, let the plugin flip response code to 200 later on.
Then KMS plugin can say "i'm up, but not ready to serve requests yet" and kube-apiserver won't crashloop because of it.

immutableT replied:
ACK on the signature, you are correct.

Here a scenario:
kube-apiserver polls kms-plugin and the plugin responds with 500 because KMS returned "Key Destroyed Error" (or other non-retriable error). Assuming we use the logic that your propose, kube-apiserver will continue to retry for the duration of the Poll timeout which seems unnecessary to me.

So I see a couple of choices here:
1. Retry until we get either true or poll timeout expires
2. Differentiate between 500(return an error) and 503 (return false, nil)

WDYT?

awly replied:
Special-casing 500 and causing it to abort retries immediately sounds fine.
I'm not sure what that achieves though.
It'll cause kube-apiserver to crash and be restarted a few seconds later anyway.

Retries will continue indefinitely (indirectly through kube-apiserver restarts) in any case.

immutableT replied:
@awly PTAL - added a case for 500.

I agree that this may not make a lot of difference. However, I don't want to make any assumptions about how kube-apiserver will be restarted (i.e. its restart policy) or whether or not kube-apiserver should crash if kms-plugin is down.
My objective is to stop polling if KMS-Plugin returned a non-retryable error.

awly replied:
SGTM, as long as there's some way for KMS plugin implementors to discover this expectation from /healthz

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +342,62 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+}
+
+func getHealthz(url string) (bool, error) {
+	var c = &http.Client{
```
`c := &http.Client{...`

awly commented on staging/src/k8s.io/apiserver/pkg/apis/config/types.go:
```diff
@@ -89,4 +89,13 @@ type KMSConfiguration struct {
+	// HealthzURL is the url on which kube-apiserver could check the health status of kms-plugin.
+	// +optional
+	HealthzURL string
+	// HealthzTimeout defines when to give-up on checking healthz.
```
Hmm, i mis-interpreted this comment originally as "timeout for each healthz call attempt", not "timeout for all healthz calls combined, before stopping retries".
Maybe rephrase a bit to make it clearer?

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +342,62 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+
+func getHealthz(url string) (bool, error) {
+	var c = &http.Client{
+		Timeout: time.Second * 3,
```
reuse `kmsPluginConnectionTimeout` ?

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +342,62 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+		klog.Infof("Successfully validated healthz status of kms-plugin on url %v", url)
+		return true, nil
+	default:
+		body, _ := ioutil.ReadAll(resp.Body)
```
check the error

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -507,3 +510,185 @@ resources:
 }
+
+func TestIsKMSPluginReady(t *testing.T) {
+	var testCases = []struct {
```
`testCases := ...`

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -507,3 +510,185 @@ resources:
+				config.HealthzURL = tt.healthzURL
+			}
+
+			r, _ := isKMSPluginReady(config)
```
check the error

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -507,3 +510,185 @@ resources:
+}
+
+func TestGetHealthz(t *testing.T) {
+	var testCases = []struct {
```
`testCases := ...`

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -507,3 +510,185 @@ resources:
+	}
+}
+
+func TestGetHealthz(t *testing.T) {
```
This test seems redundant with `TestIsKMSPluginReady`. It doesn't seem to cover any new code paths

immutableT commented:
/assign @liggitt

fedebongio commented:
/assign @logicalhan

immutableT commented:
/test pull-kubernetes-dependencies

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/types.go:
```diff
@@ -89,4 +89,22 @@ type KMSConfiguration struct {
+}
+
+// HealthzEndpoint contains information required to validate healthz of kms-plugin.
+type HealthzEndpoint struct {
```
have you thought about how (and whether) the results from this would be incorporated into the `/healthz` endpoint for the API server? Do we need any of the threshold options used for readiness/liveness probes?
https://github.com/kubernetes/kubernetes/blob/d1e828f8dfc77d1a6a8d7b0bd8ff0006f058b8fe/staging/src/k8s.io/api/core/v1/types.go#L2009-L2035

liggitt replied:
It's not clear from these docs that this is only used on startup. my initial expectation was that this would factor into apiserver `/healthz` as well (other comments/questions were based on that assumption). If the kms goes unhealthy after start, does anything notice that and flag the apiserver as unhealthy currently?

immutableT replied:
1. I think it would be great to be able to incorporate this into /healthz of kube-apiserver. Frankly, I was not able to find an example of how to do this. Could you point me to something I could use as an example.

2. I think I may have re-invented the wheel with HealthzEndpoint; Probe would have done nicely. Should I refactor to Probe? 

3. I fixed the docs by making it clear that this check only takes place during initialization of kube-apiserver.

4. When kms-plugin goes unhealthy after kube-apiserver is started, (to the best of my knowledge) this fact is not reflected in the /healthz of the kube-apiserver. 
However, we have a metric that gets incremented every time we are unable to transform - this is how one could implement alerts for this condition. Of course, writes against Secrets are no longer possible which is also reflected in some counters at the REST layer.

logicalhan replied:
What do you think about throwing this into a [postStartHook](https://github.com/kubernetes/kubernetes/blob/83f6b8869f4430d9698077ea26f98d3fbadb17e7/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L75-L106) @liggitt? It feels boot-sequence-ish to me.

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/types.go:
```diff
@@ -89,4 +89,22 @@ type KMSConfiguration struct {
+type HealthzEndpoint struct {
+	// URL is the url on which kube-apiserver could check the health status of kms-plugin.
+	URL string
+	// Timeout defines when to give-up on sending healthz probes (defaults to 90 seconds).
```
does that mean that if KMS health checks return failure for 90 seconds, the apiserver enters a permanently unhealthy state? what is the rationale for giving up on checking this independent of whatever timeout the thing monitoring apiserver health decides to use?

immutableT replied:
Yes, this PR suggests not to start kube-apiserver when kms-plugin is unhealthy.
However, this could be a config option.

My experience showed that when kube-apiserver is allowed to start in this condition, it enters the state where Service Accounts are created, but are not backed-up by Secrets - this leads to a very messy state. 
Furthermore, kube-apiserver fails eventually anyway, it just takes longer to do so.
I think once we move away from the situation where Service Accounts are typically backed-up by Secrets we could re-think this decision.

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/types.go:
```diff
@@ -89,4 +89,22 @@ type KMSConfiguration struct {
+	Timeout *metav1.Duration
+	// RequestTimeout defines timeout for an individual healthz request (defaults to 3 seconds).
+	// +optional
+	RequestTimeout *metav1.Duration
```
without success/failure thresholds, does that mean a single failure here can put the apiserver in an unhealthy state briefly?

immutableT replied:
A single 500 status from kms-plugin's healthz during the initialization of kube-apiserver will trigger an error - preventing start-up of kube-apiserver.
503 and lack of response will be retried for the duration of the timeout.

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go:
```diff
@@ -87,4 +87,22 @@ type KMSConfiguration struct {
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+	// RequestTimeout defines timeout for an individual healthz request (defaults to 3 seconds).
+	// +optional
+	RequestTimeout *metav1.Duration `json:"requesttimeout,omitempty"`
```
nit: `requestTimeout`

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go:
```diff
@@ -87,4 +87,22 @@ type KMSConfiguration struct {
+	RequestTimeout *metav1.Duration `json:"requesttimeout,omitempty"`
+	// RetryInterval defines the interval at which to retry failed healthz checks (defaults to 5 seconds).
+	// +optional
+	RetryInterval *metav1.Duration `json:"retryinterval,omitempty"`
```
nit: `retryInterval`

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -20,30 +20,47 @@ import (
+
+var (
+	errKMSPluginUnhealthy                = errors.New("kms-plugin's is unhealthy")
+	errInvalidHealthzTimeoutValue        = errors.New("value of HealthzEndpoint.Timeout should be a positive")
```
nit, use serialized field names, since that is what the user will have to use to set these values

immutableT commented:
/test pull-kubernetes-integration

sttts commented:
https://github.com/kubernetes/kubernetes/pull/78458 is adding `/readyz` support. From a brief look, this looks like it should influence readiness, not health.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -20,30 +20,47 @@ import (
+	errInvalidHealthzTimeoutValue        = errors.New("value of timeout should be a positive")
+	errInvalidHealthzRequestTimeoutValue = errors.New("value of requestTimeout should be a positive")
+	errInvalidHealthzIntervalValue       = errors.New("value of retryInterval should be a positive")
+	errMissingHealthzURL                 = errors.New("empty HealthzConfig.URL")
```
can we inline errors like these? They are not shared anywhere. Having constants for them does not make the code easier to read.

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +346,97 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+	elapsed := time.Since(start)
+	klog.Infof("wait.PollImmediate retried the condition %d times, and took overall %v", retries, elapsed)
+
+	if pollErr != nil {
```
`return pollErr == nil, nil`

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:
```diff
@@ -507,3 +512,173 @@ resources:
+		},
+		{
+			desc: "Timeouts and interval must be set to defaults",
+			in: &apiserverconfig.HealthzConfig{
```
Should this struct be empty?

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +346,97 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+		klog.Infof("KMS-Plugin's healthz return status of %d on url %v", resp.StatusCode, url)
+		return false, errKMSPluginUnhealthy
+	default:
+		klog.Infof("kms-plugin reported status: %d, msg: %s", resp.StatusCode, body)
```
nit: only log response body in Debug level.

awly commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -318,3 +346,97 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
+		return true, nil
+	// KMS-Plugin should return 500 for non-retryable errors (ex. key destroyed).
+	case http.StatusInternalServerError:
+		klog.Infof("KMS-Plugin's healthz return status of %d on url %v", resp.StatusCode, url)
```
also log the response body here

deads2k commented:
I'm not strictly against this because it is optional, but given the impact I would actually compare this unhealthiness as comparable to when a single controller in the kube-controller-manager is failing.  Because the binary is logically shared and does much more than serve a single type, being able to run and use a mostly functioning kube-apiserver is probably better than not having it at all.  In fact, you could imagine wanting the kube-apiserver up to correct the kms deployment.

If the kube-apiserver crashes when the KMS plugin is unreachable, that seems like a bug we should fix.  An inability to access secrets prevents service account token from authenticating, but with the ability to tell pods not to automount secrets I'd expect to be able to have some kinds of workloads run.  It's not as though the kube-apiserver is completely dead.

If we do go with this, I'd expect the API to be congruent to the `Probe` API used in containers for liveness and readiness.

logicalhan commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -46,6 +52,85 @@ const (
+		return nil
+	default:
+		klog.Infof("KMS-Plugin's healthz reported unhealthy status %d on url %v", resp.StatusCode, h.url())
+		return fmt.Errorf("kms-plugin reported unhealhty status: %d, msg: %s", resp.StatusCode, body)
```
s/unhealhty/unhealthy/

logicalhan commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -204,6 +206,20 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
+		}
+
+		for i, h := range healthzConfigs {
+			c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck(fmt.Sprintf("kms-plugin-%d", i), func(r *http.Request) error {
```
Is there something besides the index we can use here? If KMS-plugin has multiple probes and one of them is failing, it may be helpful on the apiserver-side to be able to distinguish the culprit.

immutableT replied:
Yes, added Name filed which uniquely identifies the corresponding KMS Provider.

immutableT commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -46,6 +48,75 @@ const (
+
+// Check encrypts and decrypts test data against KMS-Plugin's gRPC endpoint.
+func (h *KMSPluginProbe) Check() error {
+	p, err := h.Service.Encrypt([]byte("ping"))
```
this connects the KMS directly to the apiserver /healthz endpoint (which is typically allowed to be triggered unauthenticated). do we want a 1:1 relationship or do we need to make sure the KMS isn't asked to encode/decode that frequently?

immutableT replied:
I feel that rate-limiting should be the responsibility of the plugin (for example, Google's CloudKMS plugin implements client-side rate-limiting), so I think 1:1 is OK from kube-apiserver's point of view.

liggitt replied:
would you expect the plugin to rate limit by failing or blocking requests? either way, since the plugin can't distinguish between this health check and encrypt/decrypt requests made in service of secret reads/writes, it doesn't seem great to allow pings of the /healthz endpoint to interfere to that degree with KMS operations needed to service the API

immutableT replied:
You are right - the issue here is that KMS Plugin can't tell whether it is being health-checked or is servicing a valid request.

I would like to address this issue (on a separate PR) by adding support for a Health Checking Protocol for KMS-Plugin's gRPC (as was suggested by @awly  ).
See: https://github.com/grpc/grpc/blob/master/doc/health-checking.md
This way, kms-plugin can make more stringent rate-limiting decisions when receiving high volume of traffic on the Health endpoint.

From kube-apiserver's point of view, it could first try the Health endpoint and fall-back to what is currently implemented in this PR (encrypt/decrypt round-trip).

Let me know if it would make sense to add Health Checking to KMS-Plugin on this PR.

liggitt replied:
a separate PR for a dedicated KMS health check is fine, but I would rate limit how frequently we hit the KMS encrypt/decrypt endpoints for health-checking purposes in this PR

immutableT replied:
Added a rate-limiter - 1 healthz check per second.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -46,6 +48,75 @@ const (
+func (h *KMSPluginProbe) Check() error {
+	p, err := h.Service.Encrypt([]byte("ping"))
+	if err != nil {
+		klog.Warningf("Attempt to validate encrypt operation of KMS-Plugin failed with error: %v", err)
```
aren't healthcheck errors already logged?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -46,6 +48,75 @@ const (
+	}
+
+	if _, err := h.Service.Decrypt(p); err != nil {
+		klog.Warningf("Attempt to validate decrypt operation of KMS-Plugin failed with error: %v", err)
```
same question here, I didn't expect to need to log this separately

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"strings"
 	"time"
 
+	"k8s.io/apiserver/pkg/server/options/encryptionconfig"
```
nit: group with other k8s imports

immutableT commented:
@liggitt PTAL.
Removed logging statements and fixed imports.

immutableT commented:
/test pull-kubernetes-bazel-test

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -204,6 +205,20 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
+		}
+
+		for _, h := range healthzConfigs {
+			c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck(fmt.Sprintf("kms-plugin-%s", h.Name), func(r *http.Request) error {
```
this exposes the name of the configured kms in healthz output when unhealthy. are we comfortable with that, or should we just number them and include the name in the error message (which is not exposed via `/healthz`)?

immutableT replied:
I originally implemented this as a number. @logicalhan suggested to use a more descriptive name to simplify troubleshooting which makes sense to me.

liggitt replied:
including the name in the error message (which is logged and available via authenticated `/healthz/<check>` access) makes sense, but I would not expose that info via `/healthz`

immutableT replied:
Switched to using a number to differentiate between providers and added a log message that includes provider's name.

logicalhan replied:
@liggitt, why is the name of configured kms too sensitive for healthz?

immutableT replied:
The name of the KMS provider is something that is supplied by  users which, in some environments, may be considered to be PII.

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go:
```diff
@@ -87,4 +89,7 @@ type KMSConfiguration struct {
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
+	// Probe when set to true will instruct KMS Provider to setup a healthz check against KMS-Plugin.
+	// +optional
+	Probe bool `json:"probe,omitempty"`
```
if this is controlling whether we connect kms health to API server healthz, can we name this something with `healthz` in the name? If something like https://github.com/kubernetes/kubernetes/pull/78458 progresses, we'll likely want distinct control over connecting kms health to API server healthz and readyz

immutableT replied:
Renamed to InstallHealthz.

liggitt commented:
/assign @deads2k

immutableT commented:
/test pull-kubernetes-e2e-gce-100-performance
/test  pull-kubernetes-bazel-test

immutableT commented:
Added rate-limiting to protect the plugin from DoS attacks.
Removed the name of the KMS Provider from healthz check name.

@liggitt and @deads2k PTAL.

Related to this #79143

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +49,88 @@ const (
+
+// Check encrypts and decrypts test data against KMS-Plugin's gRPC endpoint.
+func (h *KMSPluginProbe) Check() error {
+	if err := h.Limiter.Wait(context.Background()); err != nil {
```
This would delay the health check.  Don't you just want to remember the last status and if you cannot immediately run the check return the last known status?

immutableT replied:
Done.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +49,88 @@ const (
+
+	for _, r := range config.Resources {
+		for _, p := range r.Providers {
+			if p.KMS != nil && p.KMS.InstallHealthz {
```
it's been a very long while since I looked in here, but I thought the API was "bytes in, bytes out".  If that happens, aren't you healthy?  If so, why is this optional and why are we building a separate client for it?

deads2k replied:
> it's been a very long while since I looked in here, but I thought the API was "bytes in, bytes out". If that happens, aren't you healthy? If so, why is this optional and why are we building a separate client for it?

Not so much against the separate construction as the factorization that I expect looks nearly identical.

immutableT replied:
I agree, there is no reason why this should be optional - now installing healthz check per KMS Provider.

Regarding the multiple connections, I think having multiple connections (with different dial options) may become useful in the future. For example, we may decide to FailFast in healthz. Having said that, I am open to refactoring this to a single connection.

immutableT commented:
/test pull-kubernetes-e2e-gce-100-performance

liggitt commented on test/integration/master/transformation_testcase.go:
```diff
@@ -90,10 +93,6 @@ func newTransformTest(l kubeapiservertesting.Logger, transformerConfigYAML strin
 		return nil, err
 	}
 
-	if e.secret, err = e.createSecret(testSecret, e.ns.Name); err != nil {
```
was this intentionally dropped?

immutableT replied:
Yes, it was.
I moved the responsibility of creating a secret under test to Tests.
That is to accommodate testing of scenarios where kube-apiserver/kms-plugin are intentionally configured to be broken - testing of healthz negative test.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +46,102 @@ const (
+	}
+
+	h.lastResponse = &kmsPluginHealthzResponse{err: nil, received: time.Now()}
+	klog.V(4).Infof("Healthz check for KMS provider: %s succeed", h.name)
```
this seems more like v5, if at all

immutableT replied:
Removed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +46,102 @@ const (
+	p, err := h.Service.Encrypt([]byte("ping"))
+	if err != nil {
+		h.lastResponse = &kmsPluginHealthzResponse{err: err, received: time.Now()}
+		klog.Warningf("Healthz check for KMS provider %s failed, error: %v", h.name, err)
```
we should include the name in the returned error so it is surfaced in healthz details. if we do that, we don't need to separately log the error here. same comment applies to the decrypt call on line 136

immutableT replied:
Done, included provider's name into the error message.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:
```diff
@@ -204,6 +205,20 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
+
+		for i, h := range healthzConfigs {
+			c.HealthzChecks = append(c.HealthzChecks, healthz.NamedCheck(fmt.Sprintf("kms-plugin-%d", i), func(r *http.Request) error {
+				return h.Check()
```
the `h` captured here by the closure is incorrect, it means if you have multiple checks, they will all test the last iterated config

liggitt replied:
add a test with at least two configs and verify the checks exercise the correct KMS

liggitt replied:
isn't the first KMS in the list used to encrypt, and subsequent ones to decrypt? would we ever encounter a case where a later KMS would only permit decrypting, so the encrypt("ping") check would fail?

immutableT replied:
Fixed the closure bug.
Added integration test for multiple providers' scenario (see, kms_treansformer_test.go TestHealthz.

Yes, theoretically, the scenario you described in the third comment is possible. KMS Administrator may remove "encrypt" IAM privilege from the service account under which the kms-plugin runs once the corresponding provider is moved into the second place - used only for decryption.
In practice, this is unlikely, but we should probably document this.
Also, there is no generic way to test decryption without providing a valid encryption payload.
I think the long term solution here is to move towards gRPC Health Checking Protocol
https://github.com/grpc/grpc/blob/master/doc/health-checking.md
KMS-Plugin developers may have vendor specific way to ascertain the status of the plugin based on the requested operation type.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +46,102 @@ const (
+	defer h.l.Unlock()
+
+	if (time.Now().Sub(h.lastResponse.received)) < kmsPluginHealthzTTL {
+		klog.V(4).Infof("Returning cached value:%v for kms-plugin's healthz response, which was received at %v", h.lastResponse.err, h.lastResponse.received)
```
I wouldn't expect a log here, or if we do, at v(5)

immutableT replied:
Removed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +46,102 @@ const (
+		klog.V(4).Infof("Returning cached value:%v for kms-plugin's healthz response, which was received at %v", h.lastResponse.err, h.lastResponse.received)
+		return h.lastResponse.err
+	}
+	klog.V(4).Info("TTL for kms-plugin's healthz expired, attempting a new check.")
```
no logging here?

immutableT replied:
Removed.

immutableT commented:
/test pull-kubernetes-integration

liggitt commented:
looks relevant
```
E0702 00:03:19.695781  107386 grpc_service.go:71] failed to create connection to unix socket: @kms-provider-2.sock, error: dial unix @kms-provider-2.sock: connect: connection refused
W0702 00:03:19.695822  107386 clientconn.go:1251] grpc: addrConn.createTransport failed to connect to {@kms-provider-2.sock 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial unix @kms-provider-2.sock: connect: connection refused". Reconnecting...
panic: Conflicting storage tracking
```

immutableT commented:
/test pull-kubernetes-integration
/test pull-kubernetes-bazel-build

immutableT commented:
@liggitt Yes, it was relevant - I was missing a call to cleanup Test API Server - fixed.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -44,8 +48,114 @@ const (
+	}
+
+	h.lastResponse = &kmsPluginHealthzResponse{err: nil, received: time.Now()}
+	klog.V(5).Infof("Healthz check for KMS provider: %s succeed", h.name)
```
do we log on other health check success? wondering when this would be useful

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/etcd_test.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"time"
 
+	"k8s.io/apiserver/pkg/server"
+
```
group imports together

liggitt commented on test/integration/master/kms_plugin_mock.go:
```diff
@@ -22,7 +22,10 @@ import (
 	"encoding/base64"
 	"fmt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
```
group google.golang.org imports together

liggitt reviewed (commented):
/approve

go ahead and squash commits, and fix a couple import and logging nits while you're at it, then lgtm

liggitt commented:
/lgtm
/approve
